src/lein_monolith/plugin.clj

```diff
@@ -188,36 +181,80 @@
 ;; ## Merged Profiles (`with-all`) Creation
 
 
-(def ^:private path-keys
+(defn- subproject-dependency
+  "Given a map entry of an internal project, return the dependency coordinates
+  for the subproject."
+  [entry]
+  [(key entry) (:version (val entry))])
+
+
+(defn- subproject-dependencies
+  "Given a map of internal projects, return a vector of dependency coordinates
+  for the subprojects."
+  [subprojects]
+  (mapv subproject-dependency subprojects))
+
+
+(def ^:private paths-keys
   "Project map keys for (re)source and test paths."
   #{:resource-paths :source-paths :test-paths})
 
 
-(defn- add-profile-paths
-  "Update a profile paths entry by adding the paths from the given project.
+(defn- add-project-paths
+  ;; Applies a set union to the current set of profile paths and any incoming
+  ;; project paths to ensure complete, unique paths.
+  [profile-paths project-paths]
+  (into (set profile-paths) project-paths))
+
+
+(defn- merge-profile-paths
+  "Update profile paths entries by adding the paths from the given project.
   Returns the updated profile."
-  [project profile k]
-  (update profile k (fn combine-colls
-                      [coll]
-                      (-> coll
-                          set
-                          (into (get project k))
-                          (vary-meta assoc :replace true)))))
+  [profile project]
+  (reduce (fn update-profile-paths
+            [profile paths-key]
+            (update profile paths-key add-project-paths (paths-key project)))
+          profile paths-keys))
+
+
+(defn- finalize-paths
+  "Sorts the paths set stored under a given key and marks the result with
+  ^:replace so Leiningen's meta-merge logic prefers the merged profile paths."
+  [profile key]
+  (with-meta (update profile key sort) {:replace true}))
+
+
+(defn- prep-subproject-for-merging
+  "Ensures that the profiles that are active in the given monolith project are
+  active in the given subproject, then activates the inherited profiles and
+  absolutizes any paths in the subproject that may be relative."
+  [monolith [_project-name subproject]]
+  (-> subproject
+      (project/project-with-profiles)
```

Comment:
This step unfortunately does a lot of file system access on Leiningen < 2.9.2: https://github.com/technomancy/leiningen/blob/f3864bc35d8280a8461b7f6c593b1919a75bef7f/leiningen-core/src/leiningen/core/project.clj#L871

On >= 2.9.2, the function has a second arity that allows for profiles to be passed in:
https://github.com/technomancy/leiningen/blob/df32368a84c3441b59d0310a991a3e10c91b28b8/leiningen-core/src/leiningen/core/project.clj#L878